acceptance-tests/src/test/java/it/com/atlassian/bitbucket/jenkins/internal/SmokeTest.java

```diff
@@ -0,0 +1,444 @@
+package it.com.atlassian.bitbucket.jenkins.internal;
+
+import io.restassured.RestAssured;
+import io.restassured.path.json.JsonPath;
+import io.restassured.response.Response;
+import io.restassured.specification.RequestSpecification;
+import it.com.atlassian.bitbucket.jenkins.internal.pageobjects.*;
+import it.com.atlassian.bitbucket.jenkins.internal.util.BitbucketUtils.*;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.lib.TextProgressMonitor;
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.eclipse.jgit.transport.CredentialsProvider;
+import org.eclipse.jgit.transport.UsernamePasswordCredentialsProvider;
+import org.hamcrest.Description;
+import org.hamcrest.TypeSafeDiagnosingMatcher;
+import org.jenkinsci.test.acceptance.junit.AbstractJUnitTest;
+import org.jenkinsci.test.acceptance.junit.WithPlugins;
+import org.jenkinsci.test.acceptance.plugins.credentials.CredentialsPage;
+import org.jenkinsci.test.acceptance.plugins.credentials.UserPwdCredential;
+import org.jenkinsci.test.acceptance.po.Build;
+import org.jenkinsci.test.acceptance.po.FreeStyleJob;
+import org.jenkinsci.test.acceptance.po.JenkinsConfig;
+import org.jenkinsci.test.acceptance.po.WorkflowJob;
+import org.junit.Before;
+import org.junit.Ignore;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+import java.net.URLEncoder;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static io.restassured.http.ContentType.JSON;
+import static it.com.atlassian.bitbucket.jenkins.internal.util.BitbucketUtils.*;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.time.Duration.ofMinutes;
+import static java.util.UUID.randomUUID;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.jenkinsci.test.acceptance.plugins.credentials.ManagedCredentials.DEFAULT_DOMAIN;
+import static org.jenkinsci.test.acceptance.po.Build.Result.SUCCESS;
+
+@WithPlugins({"mailer", "atlassian-bitbucket-server-integration"})
+public class SmokeTest extends AbstractJUnitTest {
+
+    @Rule
+    public final TemporaryFolder tempFolder = new TemporaryFolder();
+
+    private String bbsAdminCredsId;
+    private BitbucketRepository forkRepo;
+    private String serverId;
+
+    @Before
+    public void setUp() {
```

Comment:
Is there a reason why there is no `after` method. There is lots of resources being created here.

Reply:
At my last job, each test made unique test data (you'd tell the test to make user "bob" and the test would create "bob-af68d7" - the tests always referred to the data with an alias). As a result, there was no need to delete anything. One nice benefit is that you could hop into the test instance and play around with the actual user/data that the test used to figure out why a test had failed.

We delete everything in BbS tests because lots of tests use the same data (user names, project keys, etc) and _not_ deleting them would cause later test failures - so cleaning up is a requirement.

It seems like that requirement doesn't exist here. Are there stylistic reasons why you think we should still clean up?

Reply:
Yes it helps in 

1. Rerunning tests again and again on local machine. I have noticed in other ITs that if you don't clean up, test start to fail.
2. I foresee some time in future to have one Jenkins instance, one BBS instance and all of the test using these. This speeds things up hugely. We will segregate resources using unique namespace but it doesn't help in debugging if other tests resources are left.
3. I have also seen that in some places the test don't clean up but there were base infrastructure classes which keep tracks of things created and deletes them after test (or after class).  I think Jira tests does that.
 4. Last but not least, we need to have empirical evidence of knowing that clean up is a problem. For example, operations are slow or it reduces dev speed. In the absence of that, best practice say that you need to cleanup after.

Reply:
If this test leaves stuff around that causes tests to fail (either on first run or on re-run), that's definitely a problem. Either the test needs to generate totally unique data and we need to _know_ this test is re-runnable or the data needs to be deleted.

There are definitely things that need to be considered if we don't clean up test data afterwards. And I'm guessing we haven't done the due diligence, so probably best to clean up.

But... I have seen other ways of doing func testing that have a number of advantages over creating non-unique data and deleting it in every test. And they challenge some of the points you're making here about best practice, dev speed and debuggability (as evidence, [LMAX won an award for their testing shortly after I left](https://www.lmax.com/press-centre/lmax-exchange-named-best-overall-testing-project)).

Regarding point 2, LMAX's func tests ran _in parallel_ against a single running instance. That's one reason why they were blazing fast (10-15 minutes to run tens of thousands of browser/REST func tests) and - since you never knew which tests were also running on the instance - why they needed to use unique data.

For point 4, the fact that you're here saying "write more code to clean up manually" seems like empirical evidence of decreased dev speed.

Adrian, an ex-colleague, has [a bunch of blog posts about testing at LMAX](https://www.symphonious.net/testing-at-lmax/) that might be interesting reading.

Reply:
This test definitely creates things with unique IDs (and/or slugs e.g. in case of fork repos). If the ITs fail after repeated runs on the same instance, we need to fix those, but I can assure you that I've been running these tests against a single/external Jenkins and BB instance and they never fail after repeated runs.
I'll clean BB Personal tokens because of the limit Daniel mentioned, but for the rest there is no need currently, unless someone has really strong feelings about it and can convince me otherwise. :-)